Content.Server/Traits/TraitSystem.Functions.cs

```diff
@@ -0,0 +1,280 @@
+    public List<string>? PsionicPowers { get; private set; } = default!;
+
+    public override void OnPlayerSpawn(EntityUid uid,
+        IComponentFactory factory,
+        IEntityManager entityManager,
+        ISerializationManager serializationManager)
+    {
+        var prototype = IoCManager.Resolve<IPrototypeManager>();
+        var psionic = entityManager.System<PsionicAbilitiesSystem>();
+        if (PsionicPowers is null)
+            return;
+
+        foreach (var powerProto in PsionicPowers)
+            if (prototype.TryIndex<PsionicPowerPrototype>(powerProto, out var psionicPower))
+                psionic.InitializePsionicPower(uid, psionicPower, false);
+    }
+}
+
+/// Handles all modification of Known Languages. Removes languages before adding them.
+[UsedImplicitly]
+public sealed partial class TraitModifyLanguages : TraitFunction
+{
+    /// The list of all Spoken Languages that this trait adds.
+    [DataField, AlwaysPushInheritance]
+    public List<string>? LanguagesSpoken { get; private set; } = default!;
+
+    /// The list of all Understood Languages that this trait adds.
+    [DataField, AlwaysPushInheritance]
+    public List<string>? LanguagesUnderstood { get; private set; } = default!;
+
+    /// The list of all Spoken Languages that this trait removes.
+    [DataField, AlwaysPushInheritance]
+    public List<string>? RemoveLanguagesSpoken { get; private set; } = default!;
+
+    /// The list of all Understood Languages that this trait removes.
+    [DataField, AlwaysPushInheritance]
+    public List<string>? RemoveLanguagesUnderstood { get; private set; } = default!;
+
+    public override void OnPlayerSpawn(EntityUid uid,
+        IComponentFactory factory,
+        IEntityManager entityManager,
+        ISerializationManager serializationManager)
+    {
+        var language = entityManager.System<LanguageSystem>();
+
+        if (RemoveLanguagesSpoken is not null)
+            foreach (var lang in RemoveLanguagesSpoken)
+                language.RemoveLanguage(uid, lang, true, false);
+
+        if (RemoveLanguagesUnderstood is not null)
+            foreach (var lang in RemoveLanguagesUnderstood)
+                language.RemoveLanguage(uid, lang, false, true);
+
+        if (LanguagesSpoken is not null)
+            foreach (var lang in LanguagesSpoken)
+                language.AddLanguage(uid, lang, true, false);
+
+        if (LanguagesUnderstood is not null)
+            foreach (var lang in LanguagesUnderstood)
+                language.AddLanguage(uid, lang, false, true);
+    }
+}
+
+/// Handles adding Moodlets to a player character upon spawning in. Typically used for permanent moodlets or drug addictions.
+[UsedImplicitly]
+public sealed partial class TraitAddMoodlets : TraitFunction
+{
+    /// The list of all Moodlets that this trait adds.
+    [DataField, AlwaysPushInheritance]
+    public List<ProtoId<MoodEffectPrototype>>? MoodEffects { get; private set; } = default!;
```

Comment:
once again in no world it should be null
```suggestion
    [DataField(required: true), AlwaysPushInheritance]
    public List<ProtoId<MoodEffectPrototype>> MoodEffects { get; private set; } = new ();
```